python/cudf/cudf/core/column/categorical.py

```diff
@@ -56,18 +56,42 @@ def as_unordered(self, **kwargs):
 
     def add_categories(self, new_categories, **kwargs):
         new_categories = column.as_column(new_categories)
-        new_categories = self._column.categories.append(new_categories)
+        old_categories = self._column.categories
+
+        if (
+            old_categories.dtype == "object"
+            and new_categories.dtype != "object"
+        ) or (
+            old_categories.dtype != "object"
+            and new_categories.dtype == "object"
```

Comment:
Better opportunity to short circuit in the case of all numeric:
```suggestion
           new_categories.dtype == "object"
            and old_categories.dtype != "object"
```